### ISSUE TITLE
Help on using this

### ISSUE THREAD

**Issue by Danie10** [StreamController/Requests#1]
I'm trying to read the state value of a sensor in my Home Assistant. Although I'm using an auth token which works when I use curl on CLI, it is failing in StreamController as error 401 unauthorized.

So using curl like this:
```
curl -X GET http://mydomain:8123/api/states/sensor.battery_soc -H 'Authorization: Bearer tokenhere'
```

I get this returned to the CLI"
```
{"entity_id":"sensor.battery_soc","state":"54","attributes":{"unit_of_measurement":"%","friendly_name":"Battery SOC"},"last_changed":"2024-04-16T19:01:53.456290+00:00","last_reported":"2024-04-16T19:06:08.626730+00:00","last_updated":"2024-04-16T19:01:53.456290+00:00","context":{"id":"01HVM4JYNGE35JAF80AZXSG68N","parent_id":null,"user_id":null}}
```

Although it works from the CLI it does not work from my browser and also returns the same error, so it may well be the syntax or formatting needs to be adapted for the http get? I'd probably also need to parse the returned data to retrieve the state value.

But I'm just wondering if there are any examples of how we should use the get and post request?

**Comment by Core447** [comment]
I'll need to look at this tomorrow and come back to you.

But even if you get it working, the Request action won't be able to show the values retrieved by the request. Maybe I can add an option to the RunCommand action that shows the return value of a command. This would allow displaying your sensor values.

**Comment by Danie10** [comment]
OK thanks no rush - would just like to know otherwise what and how it is used. If it could retrieve the JSON data, that could technically be parsed to show a value much like we get with the CPU% in a button under OS plugin.

**Comment by Core447** [comment]
I added an option to the Run Command action to show the result of the command. You can use this as a workaround until I implement something to display retrieved values in the Requests plugin.

**Comment by Danie10** [comment]
OK thanks. Got the update in the Store, but it was not showing anything after a button press. I'm temporarily using this URL as a test as no authorization required - http://date.jsontest.com/.

**Comment by Core447** [comment]
The command `sh -c "curl -s http://date.jsontest.com/ | jq -r '.date'"` works fore me. The flatpak version automatically adds `flatpak-spawn --host` to the first command, this is why the `sh -c` stuff is necessary.

**Comment by Danie10** [comment]
OK got you the Flatpak, yes using Flatpak version. For some reason, still not returning anything to the button. I briefly see a triangle warning and then it goes blank. I also tried with `bash -c` which I used under the old software. I also tried chnaging the background colour of the button in case it was same colour.

If I run your command on command line, it does execute and returns the date.

Are you seeing the date return in the button?

**Comment by Core447** [comment]
Yes, ˋsh -c "curl -s http://date.jsontest.com/ | jq -r '.date'"ˋ works for me. If you run StreamController from the command line you should be able to see error messages

**Comment by Danie10** [comment]
OK started seeing the connection failed messages below as soon as I opened that page that has the http request on. The error came out after pressing the button:

```
2024-04-18 14:14:14.618 | INFO     | __main__:update_assets:136 - Updating 0 store assets took 9.473981618881226 seconds
Connection failed, reconnecting in 5 second(s).
Connection failed, reconnecting in 5 second(s).
Connection failed, reconnecting in 5 second(s).
Connection failed, reconnecting in 5 second(s).
Connection failed, reconnecting in 5 second(s).
Connection failed, reconnecting in 5 second(s).
2024-04-18 14:14:43.526 | ERROR    | plugins.com_core447_Requests.main:on_key_down:113 - No connection adapters were found for 'sh -c "curl -s http://date.jsontest.com/ | jq -r \'.time\'"'
Connection failed, reconnecting in 5 second(s).
```

Interestingly, though, as I go back to other pages the connection failed still repeats without any keys pressed. Deck is responding fine so not sure what connection is keeps retrying (maybe the http get one?).

**Comment by Danie10** [comment]
Wow OK works 100% and is returning just that state value for battery state of charge percentage in the button now from my Home Assistant. So right now it is press and get update. I suppose what would be very useful on this command is an auto update interval that one can set for 60 secs or 300 secs etc. Then it will show the updated value whenever that page is in focus.

**Comment by Core447** [comment]
I'll add an auto update to the requests plugin when adding support for displaying the results

**Comment by Danie10** [comment]
OK great thanks very much for helping me get this going here.

**Comment by Core447** [comment]
I added new options to the Get Request action that should allow you to use it instead of the run command one

**Comment by Danie10** [comment]
Thanks I see it now - how do we structure it with the keys separately now?


Sent with Proton Mail secure email.

**Comment by Core447** [comment]
You can use periods to separate different keys

**Comment by Danie10** [comment]
OK sorry, maybe you can help expand a bit as what I tried did not seem to work. This is my current run command:
```
sh -c "curl -X GET http://192.168.1.212:8123/api/states/sensor.battery_soc -H 'Authorization: Bearer token' | jq -r '.state'"
```

**Comment by Core447** [comment]
In order to be able to use the Get action, you'll need a single url. You can then enter "state" as a key.

I believe this url should work: `http://mydomain:8123/api/states/sensor.battery_soc?Authorization=token`

**Comment by Danie10** [comment]
Nope, I'm still having some trouble. It is hitting my HA server, but HA is expecting to see word Bearer in, I think as part of the header passed.

I'm currently trying:
```
http://192.168.1.212:8123/api/states/sensor.battery_soc?Authorization="Bearer token"
```

I get error back:
```
2024-04-20 15:56:20.682 | ERROR    | plugins.com_core447_Requests.main:on_key_down:137 - Extra data: line 1 column 4 (char 3)
```

**Comment by Core447** [comment]
Can you see some json if you open the URL in a browser?

**Comment by Danie10** [comment]
It's giving me unauthorized access right now. So the syntax is not right going to Home Assistant. It wants the header passed like this from CLI: ` -H 'Authorization: Bearer tokenhere'`

**Comment by Danie10** [comment]
I'm reading through this at https://developers.home-assistant.io/docs/api/rest/ to see if there is further guidance.

**Comment by Core447** [comment]
I updated the request action. You should now be able to add headers

**Comment by Danie10** [comment]
OK tried:
```
http://192.168.1.212:8123/api/states/sensor.battery_soc?Authorization="Bearer tokenhere"
```
Gives:
```
2024-04-20 16:26:48.464 | ERROR    | plugins.com_core447_Requests.main:on_key_down:137 - Extra data: line 1 column 4 (char 3)
```

And:
```
http://192.168.1.212:8123/api/states/sensor.battery_soc -H 'Authorization: Bearer tokenhere'
```
Gives:
```
2024-04-20 16:28:45.538 | ERROR    | plugins.com_core447_Requests.main:on_key_down:137 - Extra data: line 1 column 4 (char 3)
```

**Comment by Danie10** [comment]
Sorry wait I needed to update the plugin

**Comment by Danie10** [comment]
OK still getting an error:
```
2024-04-20 16:38:23.878 | ERROR    | plugins.com_core447_Requests.main:on_key_down:153 - 'str' object has no attribute 'items'
```

Am using:
URL: ```http://192.168.1.212:8123/api/states/sensor.battery_soc```
Header: ```{"Authorization": "Bearer tokenhere", "content-type": "application/json",}```

**Comment by Core447** [comment]
I made a mistake with the headers, sry. Can you please update and try again?

**Comment by Danie10** [comment]
```
2024-04-20 16:49:11.984 | ERROR    | plugins.com_core447_Requests.main:on_key_down:153 - Expecting property name enclosed in double quotes: line 1 column 248 (char 247)
```

**Comment by Core447** [comment]
Can you try to set the keys to `state`

**Comment by Danie10** [comment]
OK I tried it with plain state nothing else:

```
2024-04-20 16:51:19.142 | ERROR    | plugins.com_core447_Requests.main:on_key_down:153 - Expecting property name enclosed in double quotes: line 1 column 248 (char 247)
```

**Comment by Danie10** [comment]
Think the error is in header field maybe?

**Comment by Core447** [comment]
> Header: `{"Authorization": "Bearer tokenhere", "content-type": "application/json",}`

Can you try again without the last comma? -> `{"Authorization": "Bearer tokenhere", "content-type": "application/json"}`

**Comment by Danie10** [comment]
Bingo!!! Works. I had tried taking it out earlier, but obviously it also needed that last fix you did.

**Comment by Core447** [comment]
Nice, sorry that this took so long

**Comment by Danie10** [comment]
Thanks very much. I'll test out the auto fetch a bit later, and something with two keys. Hopefully it is generic enough for everyone else to still use.

No problem, I really appreciate getting it right. Want to do a video in about a week showing some of what can be done now with this, also for non-OBS users.

**Comment by Core447** [comment]
The headers are still quite generic. I'll add them to the post request method in the next days.

The video sounds really nice! I would appreciate it if you could share it on the [Discord](https://discord.gg/MSyHM8TN3u) if you're done.

**Comment by Danie10** [comment]
OK I'll certainly do so - going to try make it appeal as broadly as possible. The app will be developing all the time, so this a would really be an appetite whetter for people to consider Stream Deck on Linux for all sorts of uses. The app is already at a very impressive level I think. Idea would also be, with this baseline established, to try to attract more plugin etc devs.

**Comment by Core447** [comment]
Thank you!
